snapcraft/parts/plugins/initrd_build.sh

```diff
@@ -0,0 +1,503 @@
+      '*' \
+      "${chroot_home}/usr/lib/ubuntu-core-initramfs/main"
+  done
+  unset IFS
+
+  rm -rf "${chroot_home}/boot/initrd"*
+
+  run_chroot "$chroot_home"              \
+    "ubuntu-core-initramfs create-initrd \
+      --kernelver \"$kver\"              \
+      --output /boot/initrd.img"
+
+  install -Dm644 "${chroot_home}/boot/initrd.img-${kver}" \
+    "$CRAFT_PART_INSTALL/initrd.img-${kver}"
+
+  ln -sf "initrd.img-${kver}" "${CRAFT_PART_INSTALL}/initrd.img"
+}
+
+prep_sign() {
+  chroot_home="$1"
+  key="$2"
+  cert="$3"
+
+  # Use snakeoil keys  by default
+  key="${chroot_home}/${key}"
+  cert="${chroot_home}/${cert}"
+
+  for dir in "$CRAFT_STAGE" "$CRAFT_PROJECT_DIR"; do
+    if [ -e "${dir}/${key}" ]; then
+      printf 'Using %s in %s\n' "$key" "$dir"
+      key="${dir}/${key}"
+    fi
+
+    if [ -e "${dir}/${cert}" ]; then
+      printf 'Using %s in %s\n' "$cert" "$dir"
+      cert="${dir}/${cert}"
+    fi
+  done || printf 'Using snakoil key and cert\n'
+
+  # Ensure chroot has been setup
+  [ -e "$base_created"    ] || setup_chroot     "$chroot_home"
+  [ -e "$base_configured" ] || configure_chroot "$chroot_home"
+
+  cp --link "$key"  "${chroot_home}/root/${key##*/}"
+  cp --link "$cert" "${chroot_home}/root/${cert##*/}"
+}
+
+create_efi() {
+  chroot_home="$1"
+  key="$2"
+  cert="$3"
+
+  # Ensure chroot has been setup
+  [ -e "$base_created"    ] || setup_chroot     "$chroot_home"
+  [ -e "$base_configured" ] || configure_chroot "$chroot_home"
+
+  rm -f                             "${chroot_home}/boot/kernel.efi"*
+  ln -f "${CRAFT_STAGE}/kernel.img" "${chroot_home}/boot/kernel.img-${kver}"
+
+  run_chroot "$chroot_home"             \
+    "ubuntu-core-initramfs create-efi   \
+      --kernelver \"$kver\"             \
+      --key       \"/root/${key##*/}\"  \
+      --cert      \"/root/${cert##*/}\" \
+      --initrd    /boot/initrd.img      \
+      --kernel    /boot/kernel.img      \
+      --output    /boot/kernel.efi"
+
+  # Remove unecessary initrd file
+  rm -f  "${CRAFT_PART_INSTALL}/initrd.img-"*
+  unlink "${CRAFT_PART_INSTALL}/initrd.img"
+
+  install -Dm644 "${chroot_home}/boot/kernel.efi-${kver}" \
+    "${CRAFT_PART_INSTALL}/kernel.efi-${kver}"
+
+  ln -sf "kernel.efi-${kver}" "${CRAFT_PART_INSTALL}/kernel.efi"
+}
+
+# Ensure building the initrd in a native environment, so build within a chroot
+# Build within CRAFT_PART_SRC to avoid issues related to iterative builds
+printf 'Preparing to build initrd for arch %s using series %s in %s\n' \
+  "$CRAFT_ARCH_BUILD_FOR" "$UBUNTU_SERIES" "$CRAFT_PART_SRC"
+setup_chroot "$initrd_root"
+
+# Install kernel firmware, modules into chroot
+printf 'Installing firmware and modules into chroot\n'
+rm -rf "${initrd_root}/usr/lib/firmware/"* \
+       "${initrd_root}/usr/lib/modules"/*
+cp --archive --link --force "$kernel_firmware" "${initrd_root}/usr/lib/firmware"
+cp --archive --link --force "$kernel_modules"  "${initrd_root}/usr/lib/modules"
```

Comment:
Something somewhere is causing *all* kernel modules to end up in the chroot, even though only a handful end up in the snap proper. To investigate.